### PR TITLE
Refactor all the servers/listeners

### PR DESCRIPTION
**Pull request by kshlm**
... and make use of github.com/thejerf/suture to manage them.

More information on `suture` can be found in this [blog post][1]

This PR, refactors the GD2 rest and grpc servers, to make them suture services.
A new `muxsrv` package has been created, creates a cmux server supervisor,
which currently manages the rest server.  A `servers` package implements a
supervisor which manages the above.

The `main` function creates and starts a global/main supervisor, and then adds
other services/supervisors to it.  Right now, this is just the servers
supervisor. The etcdmgmt service will be converted to be a suture service, and
be managed by the main supervisor.

[1]: http://www.jerf.org/iri/post/2930

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -79,52 +77,34 @@ func main() {
...
-				os.Exit(0)
-
-			default:
-				continue
-			}
+	for s := range sigCh {
+		log.WithField("signal", s).Debug("Signal recieved")
+		switch s {
+		case os.Interrupt:
+			log.WithField("signal", s).Info("Recieved SIGTERM. Stopping GlusterD.")
```
Comment (kshlm) [inline main.go]: I had the opposite experience. When the signal handler was in a goroutine it didn't catch signals. Having the main thread do it worked for me. I'll need to check this further.

**Comment by kshlm** [comment]
I'm still investigating the panics on shutdown.

**Comment by prashanthpai** [comment]
@kshlm  Do you mind splitting up the PR into two parts as there are 20 commits which change 40 files.

1. PR to move/reorganize code into `\servers`. And rename imports/packages, remove old dependencies. Basically everything but the suture specific changes. This can be quickly merged as it doesn't change anything functionally.
2. PR to 'suturize' servers.

If it's too much work to split it up, please let me know, and I'll review the changes as is.

**Comment by kshlm** [comment]
> @kshlm Do you mind splitting up the PR into two parts as there are 20 commits which change 40 files.
> 
> PR to move/reorganize code into \servers. And rename imports/packages, remove old dependencies. Basically everything but the suture specific changes. This can be quickly merged as it doesn't change anything functionally.
> PR to 'suturize' servers.
> If it's too much work to split it up, please let me know, and I'll review the changes as is.

I'd prefer not to do it as. The commits as I've done them don't allow easy seperation separate PRs. I would have to redo the whole branch again, losing some history in the process.

I moved the packages using the [`gomvpkg`](https://godoc.org/golang.org/x/tools/cmd/gomvpkg) tool, which does all the changes required through out the project. So you needn't worry about reviewing the move. This includes all the changes under `commands`.

I want you to just review the changes I did for `suture`.

File: servers/rest/rest.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+func (r *GDRest) Serve() {
+	log.WithField("ip:port", r.listener.Addr().String()).Info("Started GlusterD REST server")
+	http.Serve(r.listener, r.Routes)
+	return
+}
+
+// Stop stops the GlusterD Rest server
+func (r *GDRest) Stop() {
+	log.Debug("Stopping the GlusterD Rest server")
+	r.listener.Close()
```
Comment (prashanthpai) [inline servers/rest/rest.go]: I'm curious whether this Close() here affects only this cmux'd connection or the base/main listener.

File: doc/supervisor-trees.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+      +-->peerrpc
+      |
+      +-->gd2-muxserver
+          |
+          |-->muxlistener
+          |
+          |-->rest
+          |
+          +-->sunrpc
+
```
Comment (prashanthpai) [inline doc/supervisor-trees.md]: Please add a note on what order these are stopped and restarted. To-down/bottom-up ?

File: servers/sunrpc/program.go [hunk 4]
```diff
@@ -44,13 +52,20 @@ func registerProgram(server *rpc.Server, program Program, port int) error {
...
 			sunrpc.ProcedureID{
-				program.Number(),
-				program.Version(),
-				procedure.Number,
+				ProgramNumber:   program.Number(),
+				ProgramVersion:  program.Version(),
+				ProcedureNumber: procedure.Number,
 			}, typeName+"."+procedure.Name)
 		if err != nil {
+			llogger.WithError(err).Error("failed to register sunrpc procedure")
```
Comment (prashanthpai) [inline servers/sunrpc/program.go]: nit: The caller already logs this.

Note to self: Make [sunrpc.DumpProcedureRegistry()](https://godoc.org/github.com/prashanthpai/sunrpc#DumpProcedureRegistry) to take a logger/writer.

File: servers/sunrpc/server.go [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+type SunRPC struct {
+	server   *rpc.Server
+	listener net.Listener
+	stop     chan bool
+}
+
+var programsList []Program
+
+// New returns a SunRPC server configured to listen on the given listener
+func New(l net.Listener) *SunRPC {
```
Comment (prashanthpai) [inline servers/sunrpc/server.go]: Does this need to be exported ?

File: servers/sunrpc/server.go [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+		// externally for:
+		//     1. Sending notifications to glusterfs clients
+		//        (example: volfile changed)
+		//     2. Tracking number of glusterfs clients that
+		//        are connected to glusterd2.
+		// Multiple goroutines can safely invoke write on an
+		// instance of net.Conn simultaneously.
+		conn, err := s.listener.Accept()
+		if err != nil {
+			// TODO: Handle error ?
```
Comment (prashanthpai) [inline servers/sunrpc/server.go]: This should be logged, sorry, my fault.

File: servers/sunrpc/dump_prog.go [hunk 7]
```diff
@@ -1,4 +1,13 @@
...
+package sunrpc
+
+const (
+	dumpProgNum     = 123451501
+	dumpProgVersion = 1
+
+	_ = iota
+	gfDumpDump
+	gfDumpPing
+)
```
Comment (prashanthpai) [inline servers/sunrpc/dump_prog.go]: Something has changed here. I get this now:

```
2017-01-30 19:59:12.453864 I | Remote server has no such procedure: {ProgramNumber:123451501 ProgramVersion:1 ProcedureNumber:1}
INFO[0025] glusterfs client connected                    address=127.0.0.1:49149
2017-01-30 19:59:12.456761 I | Remote server has no such procedure: {ProgramNumber:123451501 ProgramVersion:1 ProcedureNumber:1}
INFO[0025] glusterfs client connected                    address=127.0.0.1:49148
```

I can take a look at this later.

File: servers/sunrpc/server.go [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+	return port
+}
+
+// Serve will start accepting Sun RPC client connections on the listener
+// provided.
+func (s *SunRPC) Serve() {
+	log.WithField("ip:port", s.listener.Addr().String()).Info("started GlusterD SunRPC server")
+	for {
+		select {
+		case <-s.stop:
```
Comment (prashanthpai) [inline servers/sunrpc/server.go]: This looks like a neat trick! I should dig deeper into channels.

File: servers/rest/rest.go [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+)
+
+// GDRest is the GlusterD Rest server
+type GDRest struct {
+	Routes   *mux.Router
+	listener net.Listener
+}
+
+// New returns a GDRest object which can listen on the configured address
+func New(l net.Listener) *GDRest {
```
Comment (kshlm) [inline servers/rest/rest.go]: If we ever want not to use the mux server, it's gonna be here.

File: servers/rest/rest.go (hunk 2 above)
Comment (kshlm) [inline servers/rest/rest.go]: It is the underlying net.Listener, the one we used to create the mux conn. So I shouldn't be closing here.

File: servers/sunrpc/dump_prog.go (hunk 7 above)
Comment (kshlm) [inline servers/sunrpc/dump_prog.go]: Something I did wrong with iota most likely.

File: servers/sunrpc/server.go (hunk 8 above)
Comment (kshlm) [inline servers/sunrpc/server.go]: This is the starting point for graceful shutdowns.

File: servers/sunrpc/server.go (hunk 6 above)
Comment (kshlm) [inline servers/sunrpc/server.go]: Will do.

File: servers/muxsrv/server.go [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+	mux.l = l
+	mux.m = cmux.New(l)
+
+	return mux
+}
+
+// Serve starts the handlers and the multiplexed listener
+func (m *muxSrv) Serve() {
+	log.Info("started muxsrv listener")
+	m.m.Serve()
```
Comment (prashanthpai) [inline servers/muxsrv/server.go]: Both `m.m.Serve()` and `m.l.Close()` can return error. It'll be good to catch and  log these as well.

File: servers/peerrpc/listener.go [hunk 11]
```diff
@@ -9,44 +10,54 @@ import (
...
+	s.server.Serve(l)
+	return
 }
 
-// StopServer stops the server
-func StopServer() {
-	server.GracefulStop()
+// Stop stops the server
+func (s *Server) Stop() {
+	s.server.GracefulStop()
```
Comment (prashanthpai) [inline servers/peerrpc/listener.go]: Okay. Just checked and `s.server.GracefulStop()` internally closes all listeners and this isn't muxed :+1:

File: servers/rest/rest.go [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+		//TODO: Correctly handle valid errors. We could also be having errors when stopping
+		log.WithError(err).Error("GlusterD ReST server failed")
+	}
+	return
+}
+
+// Stop stops the GlusterD Rest server
+func (r *GDRest) Stop() {
+	log.Debug("stopping the GlusterD Rest server")
+	if err := r.listener.Close(); err != nil {
```
Comment (prashanthpai) [inline servers/rest/rest.go]: I think you shouldn't close the listener here as it's already closed in parent mux.

**Comment by prashanthpai** [comment]
The changes look good. I quickly run a sanity check (for the lack of functional tests) and things seem to work as expected :+1: Terminating/stopping messages (all stylistic) can be improved later.

```
INFO[0039] Etcd embedded server is stopped.             
INFO[0039] Stopped GlusterD SunRPC server               
DEBU[0039] stopping muxsrv listener                     
INFO[0039] stopped muxsrv listener                      
ERRO[0039] mux listener failed                           error=accept tcp [::]:22007: use of closed network connection
ERRO[0039] failed to accept incoming connection          error=mux: listener closed
DEBU[0039] stopping the GlusterD ReST server            
INFO[0039] stopped GlusterD ReST server                 
INFO[0039] Terminating GlusterD.                        
```

This is unrelated but found when trying out this PR:
`etcdmgmt.DestroyEmbeddedEtcd` is called on `Ctrl+C` and the old on-disk etcd data/store is wiped during shutdown and on restart of glusterd2, it's a new etcd cluster! which isn't what we want. I was wondering why this wasn't hit earlier during my runs in master branch though. May be the signal handling loop was broken in master branch and this PR fixes it.

**Comment by kshlm** [comment]
> This is unrelated but found when trying out this PR:
> etcdmgmt.DestroyEmbeddedEtcd is called on Ctrl+C and the old on-disk etcd data/store is wiped during shutdown and on restart of glusterd2, it's a new etcd cluster! which isn't what we want. I was wondering why this wasn't hit earlier during my runs in master branch though. May be the signal handling loop was broken in master branch and this PR fixes it.

Signal handling was broken. We weren't cleaning up anything. For whatever reason having the signal handler in a goroutine wasn't working.

**Pull request merged**